File: scimitar-runtime/src/main/java/com/creations/scimitar_runtime/state/StateObserver.java
package com.creations.scimitar_runtime.state;

import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

public abstract class StateObserver<D> implements Observer<Resource<D>> {

    public void onSuccess(D data){}
    public void onError(Throwable throwable){}
    public void onLoading(){}

    @Override
    public final void onChanged(@Nullable Resource<D> res) {
        if (res != null) {
            if (res.success()) {
                onSuccess(res.data);
            } else if (res.error()) {
                onError(res.getError());
            } else {
                onLoading();
            }
        }
    }
}

File: scimitar-runtime/src/main/java/com/creations/scimitar_runtime/state/StateObserver.kt
package com.creations.scimitar_runtime.state

import androidx.annotation.Nullable
import androidx.lifecycle.Observer
import com.creations.scimitar_runtime.state.Status.*

abstract class StateObserver<D> : Observer<State<D>> {

    open fun onSuccess(data: D) {}
    open fun onError(error: StateError) {}
    open fun onLoading() {}
    open fun onNoResults() {}

    override fun onChanged(@Nullable state: State<D>?) {
        if (state != null) {
            when (state.status) {
                Success -> onSuccess(state.data)
                Error -> onError(state.error!!)
                NoResults -> onNoResults()
                Loading -> onLoading()
            }
        }
    }
}
